RealmSwift/Projection.swift

```diff
@@ -321,15 +344,21 @@ extension ProjectionObservable {
 
             var projectedChanges = [PropertyChange]()
             for i in 0..<newValues.count {
-                for property in schema.filter({ $0.originPropertyKeyPathString == names[i] }) {
+                for property in schema.filter({ prop in
+                    guard let keyPaths = keyPaths, !keyPaths.isEmpty else {
+                        return prop.originPropertyKeyPathString.components(separatedBy: ".").first == names[i]
+                    }
+                    // This will allow us to notify `PropertyChange`s associated only to the keyPaths passed by the user, instead of any Property which has the same root as the notified one.
+                    return prop.originPropertyKeyPathString.components(separatedBy: ".").first == names[i] && keyPaths.contains(prop.originPropertyKeyPathString) }) {
```

Comment:
This could avoid duplicating the condition by rearranging the checks slightly:

```swift
if prop.originPropertyKeyPathString.components(separatedBy: ".").first != names[i] {
    return false
}
guard let keyPaths, !keyPaths.isEmpty else {
    return true
}
// This will allow us to notify `PropertyChange`s
// associated only to the keyPaths passed by the user,
// instead of any Property which has the same root as the
// notified one.
return keyPaths.contains(prop.originPropertyKeyPathString)
```

Since we now require swift 5.7 we can just do `if let x {` rather than `if let x = x {`.

Reply:
thanks of the tip, din't know this `if let x {`